### ISSUE TITLE
Freeze some elements of a codebook

### ISSUE THREAD

**Issue by AlecMRogers** [lucidrains/vector-quantize-pytorch#210]
Id like to freeze some elements of a codebook.

* Primary use case: some vectors are learned, and I dont want them to change anymore.
* Secondary use case: the codebook spans multiple subspaces, and I want to free some rows corresponding to those subspaces accrues all vectors.

What I think would work: 
* A way to access the grad of the codebook without having to be too intrusive on the VQP codebase.

What Ive Tried:
vq.codebook.register_hook() after learnable_codebook = True and ema_update=False never gets called.
Updating self.vq.codebook.grad[self.frozen] = 0 after the call to  loss.backward() and before the call to optimizer.step() tells me that grad is not present (or that The .grad attribute of a Tensor that is not a leaf Tensor is being accessed).

Thanks for considering.

**Comment by lucidrains** [comment]
@AlecMRogers you are using EMA updates and not the learnable codebook right? (It was only for trying to hack in the feature?)

**Comment by AlecMRogers** [comment]
That's correct, the two arguments that Im providing were only to make the codebook a nn.Parameter, which seems to be happening, but I still can't register a hook that gets called.

**Comment by lucidrains** [comment]
@AlecMRogers ok! will make the per code EMA decay thing a reality in a few

**Comment by lucidrains** [comment]
@AlecMRogers ok, try the latest version, using the `ema_update_weight` kwarg on forward

**Comment by lucidrains** [comment]
@AlecMRogers also, i once had a 508 area code number
